JoodHajeer/assignment1/Q2

```diff
@@ -42,72 +44,84 @@ private:
 	SLLNode<T> * tail;
 };
 template <class T>
-SLList<T> :: SLList(){
-	head=tail=NULL;
+SLList<T> ::SLList() {
+	head = tail = nullptr;
 }
 template <class T>
-SLList<T> :: ~SLList(){
+SLList<T> ::SLList(const SLList<T> & copy_from) {
+	head = tail = nullptr;
+	if (copy_from.head == nullptr)
+		return;
+	SLLNode<T> * current_node = copy_from.head;
+	while (current_node != nullptr) {
+		this->addToTail(current_node->val);
+		current_node = current_node->next;
+	}	
+}
+template <class T>
+SLList<T> :: ~SLList() {
 	clear();
 }
 template <class T>
-bool SLList<T> :: IsEmpty(){
-	return head==NULL;
+bool SLList<T> ::IsEmpty() {
+	return head == nullptr;
 }
 template <class T>
-void SLList<T> :: clear(){
+void SLList<T> ::clear() {
 	SLLNode<T> *p;
-	p=head;
-	tail=NULL;
-	while(p!=NULL){
-		head=head->next;
+	p = head;
+	tail = nullptr;
+	while (p != nullptr) {
+		head = head->next;
 		delete p;
-		p=head;
+		p = head;
 	}
 }
 template <class T>
-void SLList<T> :: addToTail(T &VAL){
-	SLLNode<T> *p=new SLLNode<T>(VAL,NULL);
-	if(tail==NULL)
-		head=p;
+void SLList<T> ::addToTail(T &VAL) {
+	SLLNode<T> *p = new SLLNode<T>(VAL, nullptr);
+	if (tail == nullptr)
+		head = p;
 	else
-		tail->next=p;
-	tail=p;
+		tail->next = p;
+	tail = p;
 }
 template <class T>
-int SLList<T> :: size(){//computes and return the current size of the list
-	int count=0;
+int SLList<T> ::size() {//computes and return the current size of the list
+	int count = 0;
 	SLLNode<T> * p;
-	for(p=head;p!=NULL;p=p->next)//empty list will not enter this loop
+	for (p = head; p != nullptr; p = p->next)//empty list will not enter this loop
 		count++;
 	return count;
 }
 template <class T>
-T SLList<T> :: findElementToLast(int k){//assuming k belongs to the interval [0,size)
-	int sz=this->size();//gets the current size of the list
-	k=sz-k;//kth element to last is (size-k)th element to first
-	if(k<=0){
-		return -1;//handles cases when k has a wrong value or list is empty
-	}
-	SLLNode <T> * p=head;
-	for(int i=1; i<k ;i++)//searches for elemnt at position k to last
-		p=p->next;
+T SLList<T> ::findElementToLast(int k) {
+	int sz = this->size();//gets the current size of the list
+	if (k < 0 || k >= sz) //handles cases when k has a wrong value or list is empty
+		exit(1);
```

Comment:
A bit extreme for not finding a element in a list :)
Options are either return a const pointer (and nullptr for element not found), or return a bool to indicate success and have a out parameter for the node, a declaration like:
bool findElementToLast(int k, SLLNode *element)